doc/adr/0003-self-hosted.md

```diff
@@ -0,0 +1,86 @@
+# 3. Self Hosted
+
+Date: 2021-04-27
+
+## Status
+
+Proposed
+
+## Context
+
+We want to enable enterprise customers to self host Athens knowledge graph on premise.  
+Uses should be able to collaborate on same graph, edit block & pages as well as see presence of other collaborators.
+
+## Decision
+
+We need a backend server that at least talks Datalog.
+
+For 1st backend we've chosen Datahike, because it's closeness to Datascript that we already use.  
+Another possible backend could be Crux or Datomic.
+
+### Communication Layer
+
+Because we're running in a very controlled environment (as in we don't have to support Samsung Internet & FB embedded browser)  
+we can assume availability of modern Web APIs, so for client server communication we can rely purely on WebSockets.
+
+### Concurrency
+
+To avoid problems of concurrent updates, we'd like to have single writer approach.  
+This means that updates to central graph can be done synchronously.
+
+### Connecting clients
+
+When client connects we have to get it in sync with backend.  
+
+We have 2 options here:
+
+1. Stream all the updates that happened since last tx-id client has. (easier to do because we'll need tx ingest on client for normal connected work)
+2. Provide a snapshot of DB state. (more effective on busy DB)
+
+We also have to authorize user.  
+For now we can assume trust based system, so if you say you're "Plato" we believe you w/o verification.  
+Authentication is a topic that deserves ADR of it's own.
+
+### Updating sever
+
+When Socrates makes a modification to connected DB (create, update, delete)  
+this change has to send to connected backend to be transacted.
+
+Changes that Socrates made should not be transacted locally, they should only be transacted on backend.
+
+### Updating clients
+
+Whenever transaction is executed on the backend it should be broadcasted to connected clients.  
+Because of potential `:db/id` clashes we'll need to normalize transaction data for transport (datsync is a good example of it).
+
+Client on receiving transaction data, transacts it to local database, and the change is reflected in UI.
+
+### Presence
+
+We also want to communicate to Socrates what other collaborators are present in his context.
+
+To do that we want to broadcast presence related information via same WebSocket we have for DB.
+
+Presence event types:
+* connect/disconnect
+* viewing page
+* editing block
```

Comment:
Athens should be read-only when disconnected. Should not be allowed to edit any textareas. Can still navigate though I suppose.

Reply:
Dot should also be green/red for connected/disconnected, and yellow while syncing.

Reply:
Anything else? cc @shanberg

Reply:
@tangjeff0 Looks like this is mostly about the presence of other folks, not the sync status of the DB, but with regard to the presence stuff: hard to say without using it for a bit, but I know there are some changes I'd like to make in the near term to improve readability and accessibility
- Add icons to differentiate without color
- Adjust green for better visibility to red/green colorblind folks